sdk/core/System.ClientModel/src/Internal/SSE/StreamingClientResult.cs

```diff
@@ -0,0 +1,102 @@
+
+/// <summary>
+/// Represents an operation response with streaming content that can be deserialized and enumerated while the response
+/// is still being received.
+/// </summary>
+/// <typeparam name="T"> The data type representative of distinct, streamable items. </typeparam>
+internal class StreamingClientResult<T> : AsyncResultCollection<T>
+{
+    private readonly Func<Stream, CancellationToken, IAsyncEnumerator<T>> _asyncEnumeratorSourceDelegate;
+
+    // TODO: Add an overload that creates the type but delays sending the response
+    // for use with convenience methods.  Do we need to take a func in the public
+    // method or is there another way?
+
+    private StreamingClientResult(PipelineResponse response, Func<Stream, CancellationToken, IAsyncEnumerator<T>> asyncEnumeratorSourceDelegate)
+        : base(response)
+    {
+        Argument.AssertNotNull(response, nameof(response));
+
+        if (response.ContentStream is null)
+        {
+            throw new ArgumentException("Unable to create result from response with null ContentStream", nameof(response));
+        }
+
+        _asyncEnumeratorSourceDelegate = asyncEnumeratorSourceDelegate;
+    }
+
+    /// <summary>
+    /// Creates a new instance of <see cref="StreamingClientResult{T}"/> that will yield items of the specified type
+    /// <typeparamref name="T"/> as they become available via server-sent event JSON data on the available
+    /// <see cref="PipelineResponse.ContentStream"/>. This overload uses <see cref="ModelReaderWriter"/> via the
+    /// <see cref="IJsonModel{T}"/> interface and only supports single-item deserialization per server-sent event data
+    /// payload.
+    /// </summary>
+    /// <param name="response"> The base <see cref="PipelineResponse"/> for this result instance. </param>
+    /// <param name="cancellationToken">
+    /// The optional cancellation token used to control the enumeration.
+    /// </param>
+    /// <returns> A new instance of <see cref="StreamingClientResult{T}"/>. </returns>
+    public static StreamingClientResult<U> CreateStreaming<U>(PipelineResponse response, CancellationToken cancellationToken = default)
+        where U : IJsonModel<U>
+    {
+        return new(response, GetServerSentEventDeserializationEnumerator<U>);
+    }
+
+    public static StreamingClientResult<BinaryData> CreateStreaming(PipelineResponse response, CancellationToken cancellationToken = default)
+    {
+        return new(response, GetBinaryDataEnumerator);
+    }
+
+    private static async IAsyncEnumerator<BinaryData> GetBinaryDataEnumerator(Stream stream, CancellationToken cancellationToken = default)
+    {
+        // TODO: handle via DisposeAsync instead?
+
+        using ServerSentEventReader reader = new(stream);
+        using AsyncServerSentEventEnumerator sseEnumerator = new(reader, cancellationToken);
+        while (await sseEnumerator.MoveNextAsync().ConfigureAwait(false))
+        {
+            // TODO: more efficient way?
```

Comment:
ultimately, we need to have UTF8 StreamReader to make it efficient. But you can make it more efficient by encoding Span<char> to UTF8, not char[], i.e. don't call ToArray on the chars. Probably that does not work in netstandard 2.0 though, so we would have to have slower implementation for NS2 and faster for 6.0+. Of course, don't worry about it short term. Longer term, we should probably go all the way to Utf8StreamReader. Maybe we can ask the BCL team to add it.

Reply:
Per offline discussion, we'll attempt to move to UTF-8 only as schedules allow, but stick with StreamReader conversion to UTF-16 for now in the interest of expediency.  Please flag any opportunities you spot to improve within the context of this approach.